src/plugins/intel_gpu/src/graph/primitive_inst.cpp

```diff
@@ -124,30 +127,79 @@ void primitive_inst::update_shape() {
     GPU_DEBUG_IF(debug_config->verbose >= 4) {
         GPU_DEBUG_COUT << id() << " update shape: was: " << out_layout_str << " now: " << new_layout.to_string() << std::endl;
     }
+
     if (out_layout != new_layout)
         set_shape_change();
+
     // TODO: Get rid of this const_cast
     const_cast<program_node&>(_node).set_output_layout(new_layout);
 }
 
 void primitive_inst::realloc_if_needed() {
     GPU_DEBUG_GET_INSTANCE(debug_config);
 
-    if (!_output  || _output->get_layout().count() < _node.get_output_layout().count()) {
+    if (!_output
+        || ((_output->get_layout().count() < _node.get_output_layout().count())
+        && (max_output_layout_size < _node.get_output_layout().count()))) {
         GPU_DEBUG_IF(debug_config->verbose >= 4) {
             GPU_DEBUG_COUT << "realloc memory for node: " << id() << std::endl;
         }
         _output = allocate_output();
     } else {
         _output = _network.get_engine().reinterpret_buffer(*_output, _node.get_output_layout());
     }
+    max_output_layout_size = std::max(_output->get_layout().count(), max_output_layout_size);
 }
 
 void primitive_inst::update_impl() {
+    auto pid = id();
     if (!_node.is_type<data>() && !(_node.is_type<mutable_data>() && _node.get_dependencies().empty())) {
-        PRINT_TIME(_impl = std::move(_node.type()->choose_impl(_node)));
-        PRINT_TIME(_network.get_program()->compile());
-        PRINT_TIME(_impl->init_kernels());
+        auto get_layout_key = [&]()->std::string {
+            std::string layout_key_str = "";
+            if (!_node.is_valid_output_layout()) {
+                std::cout << id() << " has invaild output layout!" << std::endl;
+                return "";
+            }
+            layout_key_str = id() + "_" + std::to_string(_node.get_unique_id());
+
+            auto data_tensor = convert_data_tensor(_node.get_output_layout());
+            layout_key_str += "_" + kernel_selector::toString_v2(data_tensor);
+
+            for (auto in : _node.get_dependencies()) {
+                if (!in->is_constant()) {
+                    auto data_tensor = convert_data_tensor(in->get_output_layout());
+                    layout_key_str += "_" + kernel_selector::toString_v2(data_tensor);
+                }
+            }
+            return layout_key_str;
+        };
+
+        auto layout_key = get_layout_key();
+        auto ret = std::getenv("NO_USE_IMPL_CACHE");
```

Comment:
could you remove this getenv? If you need it for debug purpose, please add it to DebugConfig.

Reply:
Fixed it